File: 2018/go/src/day11/day11.go
package main

import (
	"fmt"
)

const gridSize = 301 // Adjusted for indexing 1..300

func calculatePowerLevel(x int, y int, serialNumber int) int {
	rackID := x + 10
	initial := rackID * y
	adjusted := (initial + serialNumber) * rackID
	reduced := ((adjusted / 100) % 10) - 5
	return reduced
}

func newGrid(serialNumber int) (grid [][]int) {
	grid = make([][]int, gridSize)
	for y := 1; y < gridSize; y++ {
		grid[y] = make([]int, gridSize)
		for x := 1; x < gridSize; x++ {
			grid[y][x] = calculatePowerLevel(x, y, serialNumber)
		}
	}
	return
}

func newEmptyGrid() (grid [][]int) {
	grid = make([][]int, gridSize)
	for y := range grid {
		grid[y] = make([]int, gridSize)
		for x := range grid[y] {
			grid[y][x] = 0
		}
	}
	return
}

func sumSquare(grid [][]int, sumGrid [][]int, startX int, startY int, size int) (sum int) {
	sum = sumGrid[startY][startX]
	for y := startY; y < startY+size; y++ {
		sum += grid[y][startX+size-1]
	}
	for x := startX; x < startX+size; x++ {
		sum += grid[startY+size-1][x]
	}
	sumGrid[startY][startX] = sum
	return
}

func findLargestSquare(serialNumber int, minSize int, maxSize int) (largest [4]int) {
	grid := newGrid(serialNumber)
	sumGrid := newEmptyGrid()
	for size := 1; size <= maxSize; size++ {
		for y := 1; y < gridSize-(size-1); y++ {
			for x := 1; x < gridSize-(size-1); x++ {
				newSum := sumSquare(grid, sumGrid, x, y, size)
				if size >= minSize && newSum > largest[0] {
					largest = [4]int{newSum, x, y, size}
				}
			}
		}
	}
	return
}

func main() {
	fmt.Printf("%v\n", findLargestSquare(6548, 3, 3))
	fmt.Printf("%v\n", findLargestSquare(6548, 1, 300))
}

File: python/2018/day11.py
def calcFuelCell(x, y, gridSerialNumber):
    rackID = x + 10
    powerLevelStart = (rackID * y)
    addedSerial = powerLevelStart + gridSerialNumber
    multiplyRackID = addedSerial * rackID
    reduced = (multiplyRackID // 100) % 10
    return reduced - 5

def populateGrid(gridSerialNumber):
    for y in range(1, gridSize):
        for x in range(1, gridSize):
            grid[y][x] = calcFuelCell(x, y, gridSerialNumber)

def sumMinorSquareAddition(xStart, yStart, size, sumGrid):
    sum = sumGrid[yStart][xStart]
    for i in range(size):
        sum += grid[yStart+i][xStart+size-1]
    for i in range(size):
        sum += grid[yStart+size-1][xStart+i]
    sumGrid[yStart][xStart] = sum
    return sum

def findLargest(minSize, maxSize):
    sumGrid = [[0 for x in range(gridSize)] for y in range(gridSize)]
    largest = (0, 0, 0 ,0) # sum, x, y, size
    for size in range(1, maxSize+1):
        for y in range(1, gridSize-(size-1)):
            for x in range(1, gridSize-(size-1)):
                newSum = sumMinorSquareAddition(x, y, size, sumGrid)
                if size >= minSize and newSum > largest[0]:
                    largest = (newSum, x, y, size)
    return largest

gridSize = 301 # Adjusted for indexing 1..300
grid = [[0 for x in range(gridSize)] for y in range(gridSize)]
populateGrid(6548)
print(findLargest(3, 3))
print(findLargest(1, 300))
